logicrunner/logicrunner.go

```diff
@@ -409,30 +376,28 @@ func (lr *LogicRunner) OnPulse(ctx context.Context, pulse insolar.Pulse) error {
 		if es := state.ExecutionState; es != nil {
 			es.Lock()
 
-			toSend := es.OnPulse(ctx, meNext)
+			toSend := state.ExecutionBroker.OnPulse(ctx, meNext)
 			messages = append(messages, toSend...)
 
-			if !meNext {
-				if es.Broker.currentList.Empty() {
-					state.ExecutionState = nil
-				}
-			} else {
-				if es.pending == message.NotPending && es.LedgerHasMoreRequests {
-					lr.startGetLedgerPendingRequest(ctx, es)
-				}
-				if es.pending == message.NotPending {
-					es.Broker.StartProcessorIfNeeded(ctx)
-				}
+			if !meNext && state.ExecutionBroker.currentList.Empty() {
+				// we're not executing and we have nothing to process
+				state.ExecutionState = nil
+				state.ExecutionBroker = nil
+			} else if meNext && es.pending == message.NotPending {
+				// we're executing, micro optimization, check pending
+				// status, reset ledger check status and start execution
+				state.ExecutionBroker.ResetLedgerCheck()
+				state.ExecutionBroker.StartProcessorIfNeeded(ctx)
 			}
 
 			es.Unlock()
 		}
 
-		if state.ExecutionState == nil && state.Validation == nil {
+		state.Unlock()
+
+		if state.Validation == nil && state.ExecutionState == nil {
 			lr.StateStorage.DeleteObjectState(ref)
 		}
-
-		state.Unlock()
```

Comment:
this change is invalid! if condition  ```state.Validation == nil && state.ExecutionState == nil``` should be checked under lock.

Reply:
fixed